File: Algoritmos/Biseccion.R
# Remueve todos los objetos creados
rm(list=ls())
Fx = function(x) ((x*((exp(1))^x))-pi)
Fx2 = function(x) ((exp(1)^x)-(pi*x))

biseccion = function(a,b,err) {
  # Grafico
  x = seq(a,b,0.1)
  plot(x,Fx(x),type="l",col="blue")
  abline(h=0,v=0,col="red")
  
  inf = a
  sup = b
  aux = 0
  aux2 = 0
  it = 0
  result = Fx2(b)
  
  while(err < abs(result)){
    it = it + 1
    
    if((Fx2(inf)*Fx2(sup))<0){
      aux = (sup+inf)/2
      aux2 = Fx2(aux)
      
      if(abs(aux2) < err){
        break
      }else{
        if((Fx2(inf)*aux2)>0){
          inf =  aux
        }else{
          sup = aux
        }
      }
      
    }else{
      cat("Intervalo no valido\n")
      break
    }
  }
  cat("Iteraciones =",it,"Resultado = ",aux,"\n")
}

biseccion(0,1,10e-8)
biseccion(1,2,10e-8)

File: Algoritmos/Biseccion/Biseccion.cpp
#include <iostream>
#include <cmath>

const double e = std::exp(1.0);
const double pi = M_PI;

double funcion(double x);
double biseccion(double a, double b, double err, double &it);

int main(){
	
	double result = 0;
	double aa ;
	result = biseccion(0,2,10e-8,aa);
	
    std::cout<<"Raiz: "<<result<<std::endl;
    std::cout<<"# Iteraciones: "<<aa<<std::endl;
    
    return 0;
}


double funcion(double x)
{
    return (((x)*(pow(e,x)))-pi);
}

double biseccion(double a,double b,double err, double &it){ 
	
	double inf = a;
	double sup = b;
	double aux = 0;
	double aux2 = 0;
	double ite = 0;
	double result = funcion(b);
	
	while (err < fabs(result))
	{
		ite++;
		if((funcion(inf)*funcion(sup))<0){
			aux = (sup+inf)/2;
			aux2 = funcion(aux);
			
			if(fabs(aux2) < err){
				//aux = aux2;
				break;
			}else{
				if((funcion(inf)*aux2)>0){
					inf =  aux;
				}else{
					sup = aux;
				}
			}
			
		}else{
			std::cout<<"Intervalo no valido"<<std::endl;
			return 0;
		}
	}
	it = ite;
	return aux;
}
